### PR TITLE
feat(array): generalize `**Array::from_slice` to `from_iter`

### PR DESCRIPTION
**Pull request by wangrunji0408**
Signed-off-by: Runji Wang <wangrunji0408@163.com>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

This PR generalizes array's `from_slice` function to `FromIterator` trait, so that we can build an array from iterator.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/common/src/array/primitive_array.rs [hunk 1]
```diff
@@ -132,10 +132,22 @@ pub struct PrimitiveArray<T: PrimitiveArrayItemType> {
     data: Vec<T>,
 }
 
-impl<T: PrimitiveArrayItemType> PrimitiveArray<T> {
-    pub fn from_slice(data: &[Option<T>]) -> Self {
-        let mut builder = <Self as Array>::Builder::new(data.len());
-        for i in data {
+impl<T: PrimitiveArrayItemType> FromIterator<T> for PrimitiveArray<T> {
+    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
```
Comment (BugenZhao) [inline src/common/src/array/primitive_array.rs:136]: How about forwarding the implementation to `FromIterator<&'a Option<T>>`? 🤔

**Review by BugenZhao — APPROVED** [review]
LGTM!

File: src/common/src/array/primitive_array.rs (hunk 1 above)
Comment (wangrunji0408) [reply src/common/src/array/primitive_array.rs:136]: I tried but failed due to lifetime problem. 🥹
WX20221209-131015@2x

Comment (wangrunji0408) [reply src/common/src/array/primitive_array.rs:136]: But we could forward both `FromIterator<T>` and `FromIterator<&'a Option<T>>` to `FromIterator<Option<T>>`.

**Pull request merged**